app/src/main/java/org/oppia/android/app/mydownloads/downloads/DownloadsFragmentPresenter.kt

```diff
@@ -0,0 +1,255 @@
+          /* filter =*/ false
+        )
+        downloadsRecyclerViewAdapter.setData(sortedDownloadsItemViewModel)
+      }
+    }
+  }
+
+  private fun sortTopicAlphabetically(): MutableList<DownloadsItemViewModel> {
+    val sortedTopicNameList = mutableListOf<String>()
+    val downloadsItemViewModelList = downloadsViewModel.downloadsViewModelLiveData.value
+    downloadsItemViewModelList?.forEach { downloadsItemViewModel ->
+      if (downloadsItemViewModel is DownloadsTopicViewModel) {
+        sortedTopicNameList.add(downloadsItemViewModel.topicSummary.name)
+      }
+    }
+    sortedTopicNameList.sort()
+
+    val sortedDownloadsItemViewModel = mutableListOf<DownloadsItemViewModel>()
+    sortedDownloadsItemViewModel.add(downloadsItemViewModelList!![0])
+    sortedTopicNameList.forEach { topicName ->
+      downloadsItemViewModelList.forEach { downloadsItemViewModel ->
+        if (downloadsItemViewModel is DownloadsTopicViewModel &&
+          topicName == downloadsItemViewModel.topicSummary.name
+        ) {
+          sortedDownloadsItemViewModel.add(downloadsItemViewModel)
+        }
+      }
+    }
+    return sortedDownloadsItemViewModel
+  }
+
+  // TODO: try to improvise this using kotlin functions
+  private fun sortTopicDownloadSize(): MutableList<DownloadsItemViewModel> {
+    val sortedTopicSizeList = mutableListOf<Long>()
+    val downloadsItemViewModelList = downloadsViewModel.downloadsViewModelLiveData.value
+    downloadsItemViewModelList?.forEach { downloadsItemViewModel ->
+      if (downloadsItemViewModel is DownloadsTopicViewModel) {
+        sortedTopicSizeList.add(downloadsItemViewModel.topicSummary.diskSizeBytes)
+      }
+    }
+    sortedTopicSizeList.sort()
+
+    val sortedDownloadsItemViewModel = mutableListOf<DownloadsItemViewModel>()
+    sortedDownloadsItemViewModel.add(downloadsItemViewModelList!![0])
+    sortedTopicSizeList.forEach { topicSize ->
+      downloadsItemViewModelList.forEach { downloadsItemViewModel ->
+        if (downloadsItemViewModel is DownloadsTopicViewModel &&
+          topicSize == downloadsItemViewModel.topicSummary.diskSizeBytes
+        ) {
+          sortedDownloadsItemViewModel.add(downloadsItemViewModel)
+        }
+      }
+    }
+    return sortedDownloadsItemViewModel
+  }
+
+  private fun bindDownloadsTopicCard(
+    binding: DownloadsTopicCardBinding,
+    viewModel: DownloadsTopicViewModel
+  ) {
+    binding.viewModel = viewModel
+
+    var isDeleteExpanded = false
+    binding.isDeleteExpanded = isDeleteExpanded
+
+    binding.expandListIcon.setOnClickListener {
+      isDeleteExpanded = !isDeleteExpanded
+      binding.isDeleteExpanded = isDeleteExpanded
+    }
+
+    binding.deleteImageView.setOnClickListener {
+      downloadsViewModel.profileLiveData.observe(
+        fragment,
+        Observer { profile ->
+          when {
+            profile.allowDownloadAccess -> {
+              openDownloadTopicDeleteDialog(profile.allowDownloadAccess)
```

Comment:
Keeping it open here. 

> @rt4914 bit concerned about opening dialog while observing the data. Could you suggest a good approach here?

Reply:
I don't think this should be an issue. This code does look correct to me.

Reply:
Did you find any delay when clicking on delete icon?

Reply:
No actually I am not seeing any delay.
Though technically there should be a minor delay expected because of Observer but maybe thats not significant.